flocker/control/_diffing.py

```diff
@@ -0,0 +1,247 @@
+    """
+    Computes a series of ``_DiffChange`` s to turn ``mapping_a`` into
+    ``mapping_b`` assuming that these mappings are at ``current_path`` inside a
+    nested pyrsistent object.
+
+    :param current_path: And iterable of pyrsistent object describing the path
+        inside the root pyrsistent object where the other arguments are
+        located.  See ``PMap.transform`` for the format of this sort of path.
+
+    :param mapping_a: The desired input mapping.
+
+    :param mapping_b: The desired output mapping.
+
+    :returns: An iterable of ``_DiffChange`` s that will turn ``mapping_a``
+        into ``mapping_b``.
+    """
+    resulting_diffs = pvector([]).evolver()
+    a_keys = frozenset(x for x in mapping_a.iterkeys())
+    b_keys = frozenset(x for x in mapping_b.iterkeys())
+    for key in a_keys.intersection(b_keys):
+        if mapping_a[key] != mapping_b[key]:
+            resulting_diffs.extend(
+                _create_diffs_for(
+                    current_path.append(key),
+                    mapping_a[key],
+                    mapping_b[key]
+                )
+            )
+    for key in b_keys.difference(a_keys):
+        resulting_diffs.append(
+            _Set(path=current_path.append(key), val=mapping_b[key])
+        )
+    for key in a_keys.difference(b_keys):
+        resulting_diffs.append(
+            _Remove(path=current_path.append(key))
+        )
+    return resulting_diffs.persistent()
+
+
+def _create_diffs_for(current_path, subobj_a, subobj_b):
+    """
+    Computes a series of ``_DiffChange`` s to turn ``subobj_a`` into
+    ``subobj_b`` assuming that these subobjs are at ``current_path`` inside a
+    nested pyrsistent object.
+
+    :param current_path: And iterable of pyrsistent object describing the path
+        inside the root pyrsistent object where the other arguments are
+        located.  See ``PMap.transform`` for the format of this sort of path.
+
+    :param subobj_a: The desired input sub object.
+
+    :param subobj_b: The desired output sub object.
+
+    :returns: An iterable of ``_DiffChange`` s that will turn ``subobj_a``
+        into ``subobj_b``.
+    """
+    if subobj_a == subobj_b:
+        return pvector([])
+    elif type(subobj_a) != type(subobj_b):
+        return pvector([_Set(path=current_path, val=subobj_b)])
+    elif isinstance(subobj_a, PClass) and isinstance(subobj_b, PClass):
+        a_dict = subobj_a._to_dict()
+        b_dict = subobj_b._to_dict()
+        return _create_diffs_for_mappings(current_path, a_dict, b_dict)
+    elif isinstance(subobj_a, PMap) and isinstance(subobj_b, PMap):
+        return _create_diffs_for_mappings(
+            current_path, subobj_a, subobj_b)
+    elif isinstance(subobj_a, PSet) and isinstance(subobj_b, PSet):
+        return _create_diffs_for_sets(
+            current_path, subobj_a, subobj_b)
+    return pvector([_Set(path=current_path, val=subobj_b)])
```

Comment:
:question: I'm having to think about this default. Maybe add an explanatory comment.